src/passes/InstrumentAllocations.cpp

```diff
@@ -0,0 +1,158 @@
+  Name tracerName;
+  Type params;
+  Type results;
+};
+bool operator<(const TracedFunction& lhs, const TracedFunction& rhs) {
+  return lhs.originName < rhs.originName;
+}
+
+struct InstrumentAllocations
+  : public WalkerPass<PostWalker<InstrumentAllocations>> {
+  bool addsEffects() override { return true; }
+
+  void visitCall(Call* curr) {
+    auto* target = getModule()->getFunction(curr->target);
+
+    auto function_it = std::find_if(traceFunctions.begin(),
+                                    traceFunctions.end(),
+                                    [target](const TracedFunction& f) {
+                                      return f.originName == target->name;
+                                    });
+    if (function_it != traceFunctions.end()) {
+      addInstrumentation(curr, target, function_it->tracerName);
+    }
+  }
+
+  void visitModule(Module* curr) {
+    for (const auto& traceFunction : traceFunctions) {
+      addImport(curr, traceFunction);
+    }
+  }
+
+private:
+  void addInstrumentation(Call* curr,
+                          const wasm::Function* target,
+                          const Name& wrapperName) {
+    Builder builder(*getModule());
+    std::vector<wasm::Expression*> realCallParams, trackerCallParams;
+
+    for (const auto& op : curr->operands) {
+      auto localVar = builder.addVar(getFunction(), op->type);
+      realCallParams.push_back(builder.makeLocalTee(localVar, op, op->type));
+      trackerCallParams.push_back(builder.makeLocalGet(localVar, op->type));
+    }
+
+    auto resultType = target->type.getSignature().results;
+    auto realCall = builder.makeCall(target->name, realCallParams, resultType);
+
+    if (resultType.isConcrete()) {
+      auto resultLocal = builder.addVar(getFunction(), resultType);
+      trackerCallParams.insert(
+        trackerCallParams.begin(),
+        builder.makeLocalTee(resultLocal, realCall, resultType));
+
+      replaceCurrent(builder.makeBlock(
+        {builder.makeCall(
+           wrapperName, trackerCallParams, Type::BasicType::none),
+         builder.makeLocalGet(resultLocal, resultType)}));
+    } else
+      replaceCurrent(builder.makeBlock(
+        {realCall,
+         builder.makeCall(
+           wrapperName, trackerCallParams, Type::BasicType::none)}));
+  }
+
+  void addImport(Module* wasm, const TracedFunction& f) {
+    auto import =
+      Builder::makeFunction(f.tracerName, Signature(f.params, f.results), {});
+    import->module = ENV;
+    import->base = f.tracerName;
+    wasm->addFunction(std::move(import));
+  }
+
+  std::set<TracedFunction> traceFunctions = {
+    {Name("malloc"), Name("trace_malloc"), {Type::i32, Type::i32}, Type::none},
+    {Name("calloc"),
+     Name("trace_calloc"),
+     {Type::i32, Type::i32, Type::i32},
+     Type::none},
+    {Name("realloc"),
+     Name("trace_realloc"),
+     {Type::i32, Type::i32, Type::i32},
+     Type::none},
+    {Name("free"), Name("trace_free"), {Type::i32}, Type::none}};
+};
```

Comment:
Perhaps this could be an option? I am thinking of something like
```
wasm-opt --trace-calls=malloc,i32,i32;calloc,i32,i32,i32;realloc,i32,i32,i32;free,none,i32
```
So the user provides a list of which functions to trace and their types. That would make this pretty general and maybe there are more use cases?

(For an example of a pass that gets an argument see `ExtractFunction`).

Reply:
I like the idea of having the pass more generic. I've updated the code according to your suggestion. Since the tracer receives both return value and parameters as function arguments, the pass should never expects to receive `none`; instead it expects just a sequence of (optional) return value and (optional) list of parameters.